[Fleet Monitor] petry-projects/.github-private — daily-pr-review-health.yml

## 🟠 DEGRADED: `petry-projects/.github-private` / `daily-pr-review-health.yml`

| Metric | Value |
|---|---|
| **Failure Rate** | 50% (1 / 2 runs) |
| **Status** | DEGRADED |
| **p50 Duration** | 100s |
| **p95 Duration** | 100s |
| **Successful Runs** | 1 |
| **Cancelled Runs** | 0 |

**Workflow:** [`daily-pr-review-health.yml`](https://github.com/petry-projects/.github-private/actions)

> _Last updated by [Fleet Monitor run](https://github.com/petry-projects/.github-private/actions/runs/27198381517) on 2026-06-09._
> _Threshold: failure rate > 10% over the monitored window._

## Dev-Lead Implementation Plan

**Issue:** #511 — [Fleet Monitor] petry-projects/.github-private — daily-pr-review-health.yml

### Root Cause

The `daily-pr-review-health.yml` health check fails when the Claude CLI hits its daily usage limit. From the failing run (2026-06-09T08:58:14Z):

```
##[error]Claude invocation failed. CLI output follows:
You've hit your limit · resets 4pm (UTC)
```

The script calls `exit 1` unconditionally on any Claude failure, including transient rate limits that are not indicative of a real PR review agent health problem.

### Scope

Add graceful rate-limit handling to `scripts/pr_review_health.sh` so that a transient Claude usage-cap does not mark the health-check workflow as failed. Emit a `::warning::` annotation instead and exit 0.

### Implementation Checklist
- [ ] `scripts/pr_review_health.sh` — detect rate limit in Claude error handler block; emit `::warning::`, override `HAS_FAILURES=false`, exit 0
- [ ] `tests/test_pr_review_health.bats` — new test file covering rate-limit exit-0, non-rate-limit exit-1, and no-failures skip-Claude paths
- [ ] `.github/workflows/lint.yml` — add new test file to bats test list (per AGENTS.md exception rule)

### Tests to Write
- [ ] `"Claude rate limit exits 0 and emits ::warning:: annotation"` — stub `claude` to output "You've hit your limit", verify exit 0 and `::warning::` in output
- [ ] `"Non-rate-limit Claude failure exits 1 and emits ::error:: annotation"` — stub `claude` to output generic error, verify exit 1 and `::error::` in output
- [ ] `"No failed runs exits 0 without invoking Claude"` — stub `gh` to return all-success runs, verify exit 0 and no Claude invocation

### Test Command
`bats tests/test_pr_review_health.bats`

### Pattern Reference

The rate-limit grep pattern follows `_rate_limit_pattern()` in `scripts/engine.sh` (the codebase's single source of truth), specifically the `hit your limit` and `resets [0-9]+(am|pm)` tokens that match the observed error message.

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -44,7 +44,8 @@ jobs:
             tests/test_push_protection.bats \
             tests/test_verify_auth_scopes.bats \
             tests/test_initiative_driver.bats \
-            tests/test_cut_release.bats
+            tests/test_cut_release.bats \
+            tests/test_pr_review_health.bats
 
   validate-agent-profiles:
     runs-on: ubuntu-latest

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -22,7 +22,7 @@ jobs:
       - name: shellcheck
         # --severity=warning: SC2016 (info) fires on printf '`%s`' where backticks
         # are literal markdown, not shell expressions — suppress info-level noise.
-        run: shellcheck --severity=warning scripts/fleet_monitor.sh scripts/fleet_report.sh scripts/token_report.sh scripts/lib/model-pricing.sh scripts/cut-release.sh
+        run: shellcheck --severity=warning scripts/fleet_monitor.sh scripts/fleet_report.sh scripts/token_report.sh scripts/lib/model-pricing.sh scripts/cut-release.sh scripts/pr_review_health.sh
 
   bats:
     runs-on: ubuntu-latest

--- a/scripts/pr_review_health.sh
+++ b/scripts/pr_review_health.sh
@@ -231,7 +231,7 @@ then
       "$TODAY" "$failed_runs" "$total_runs" > "$REPORT_FILE"
     # Override HAS_FAILURES so the issue-creation step does not file a spurious
     # ticket for what is a transient infrastructure constraint, not a PR review failure.
-    [ -n "${GITHUB_ENV:-}" ] && echo "HAS_FAILURES=false" >> "$GITHUB_ENV"
+    [[ -n "${GITHUB_ENV:-}" ]] && echo "HAS_FAILURES=false" >> "$GITHUB_ENV"
     rm -f "$logs_file"
     exit 0
   fi

--- a/scripts/pr_review_health.sh
+++ b/scripts/pr_review_health.sh
@@ -223,9 +223,9 @@ Output ONLY the markdown report — no preamble or commentary outside the report
 PROMPT
 then
   # Distinguish transient rate-limit/quota errors from real failures.
-  # Pattern mirrors _rate_limit_pattern() in scripts/engine.sh (the codebase's
-  # single source of truth), focused on the tokens that match Claude CLI caps.
-  if grep -qiE "hit your limit|rate.?limit|resets [0-9]+(am|pm)|usage limit|quota" "$REPORT_FILE"; then
+  # Pattern is a subset of _rate_limit_pattern() in scripts/engine.sh, focused
+  # on errors observed from the Claude CLI.
+  if grep -qiE "hit your limit|rate.?limit|resets [0-9]+(am|pm)|usage limit|quota|overloaded_error|claude.*usage" "$REPORT_FILE"; then
     echo "::warning::Claude rate limit reached — health analysis deferred. Re-run manually when the limit resets."
     printf '# PR Review Agent Health Check — %s\n\n**Analysis deferred:** Claude daily usage limit reached during log analysis. Re-run this workflow manually when the limit resets.\n\nFailed run count (unanalyzed): %d of %d total.\n' \
       "$TODAY" "$failed_runs" "$total_runs" > "$REPORT_FILE"

--- a/scripts/pr_review_health.sh
+++ b/scripts/pr_review_health.sh
@@ -222,6 +222,19 @@ Single line: \`Health: X/10 — <one-sentence verdict>\`
 Output ONLY the markdown report — no preamble or commentary outside the report sections.
 PROMPT
 then
+  # Distinguish transient rate-limit/quota errors from real failures.
+  # Pattern mirrors _rate_limit_pattern() in scripts/engine.sh (the codebase's
+  # single source of truth), focused on the tokens that match Claude CLI caps.
+  if grep -qiE "hit your limit|rate.?limit|resets [0-9]+(am|pm)|usage limit|quota" "$REPORT_FILE"; then
+    echo "::warning::Claude rate limit reached — health analysis deferred. Re-run manually when the limit resets."
+    printf '# PR Review Agent Health Check — %s\n\n**Analysis deferred:** Claude daily usage limit reached during log analysis. Re-run this workflow manually when the limit resets.\n\nFailed run count (unanalyzed): %d of %d total.\n' \
+      "$TODAY" "$failed_runs" "$total_runs" > "$REPORT_FILE"
+    # Override HAS_FAILURES so the issue-creation step does not file a spurious
+    # ticket for what is a transient infrastructure constraint, not a PR review failure.
+    [ -n "${GITHUB_ENV:-}" ] && echo "HAS_FAILURES=false" >> "$GITHUB_ENV"
+    rm -f "$logs_file"
+    exit 0
+  fi
   echo "::error::Claude invocation failed. CLI output follows:"
   cat "$REPORT_FILE" >&2
   rm -f "$logs_file"

--- a/tests/test_pr_review_health.bats
+++ b/tests/test_pr_review_health.bats
@@ -1,0 +1,174 @@
+#!/usr/bin/env bats
+# Unit tests for scripts/pr_review_health.sh
+#
+# Covers the Claude invocation error-handling paths:
+#   - Rate limit from Claude → exit 0 + ::warning:: (workflow must not fail)
+#   - Non-rate-limit Claude error → exit 1 + ::error::
+#   - No failed runs in window → exit 0 (Claude never invoked)
+#
+# Run with: bats tests/test_pr_review_health.bats
+
+SCRIPT="$(dirname "$BATS_TEST_FILENAME")/../scripts/pr_review_health.sh"
+
+setup() {
+  STUB_DIR="$(mktemp -d)"
+  TEST_WORK_DIR="$(mktemp -d)"
+  export PATH="$STUB_DIR:$PATH"
+  export GITHUB_ENV="$TEST_WORK_DIR/github.env"
+  touch "$GITHUB_ENV"
+  # Minimal required env vars
+  export LOOKBACK_DAYS=1
+  export CLAUDE_CODE_OAUTH_TOKEN=test-token-stub
+}
+
+teardown() {
+  rm -rf "$STUB_DIR" "$TEST_WORK_DIR"
+  rm -f pr_review_health_report.md
+  rm -rf health_run_logs
+}
+
+# ---------------------------------------------------------------------------
+# Helpers
+# ---------------------------------------------------------------------------
+
+# make_gh_stub: stubs gh to return controlled responses per call context.
+# Returns one failed run for the per_page=100 listing so Claude is invoked.
+make_gh_stub_one_failure() {
+  cat > "$STUB_DIR/gh" << 'STUBEOF'
+#!/usr/bin/env bash
+args="$*"
+# Token access check (per_page=1) — succeed silently
+if printf '%s' "$args" | grep -q 'per_page=1[^0]'; then
+  exit 0
+fi
+# Run listing (per_page=100) — return one failed run
+if printf '%s' "$args" | grep -q 'per_page=100'; then
+  printf '[{"id":99999,"status":"completed","conclusion":"failure","created_at":"2026-06-09T10:00:00Z","html_url":"https://github.com/petry-projects/.github-private/actions/runs/99999","run_number":99}]\n'
+  exit 0
+fi
+# Workflow source fetch (contents endpoint) — base64 "test"
+if printf '%s' "$args" | grep -q 'contents'; then
+  printf 'dGVzdA==\n'
+  exit 0
+fi
+# gh run view (log download for failed runs)
+if printf '%s' "$args" | grep -q 'run view'; then
+  printf 'step1\t2026-06-09T10:00:00Z some log line\n'
+  exit 0
+fi
+exit 0
+STUBEOF
+  chmod +x "$STUB_DIR/gh"
+}
+
+# make_gh_stub_no_failures: returns only successful runs so Claude is skipped.
+make_gh_stub_no_failures() {
+  cat > "$STUB_DIR/gh" << 'STUBEOF'
+#!/usr/bin/env bash
+args="$*"
+if printf '%s' "$args" | grep -q 'per_page=1[^0]'; then
+  exit 0
+fi
+if printf '%s' "$args" | grep -q 'per_page=100'; then
+  printf '[{"id":99998,"status":"completed","conclusion":"success","created_at":"2026-06-09T06:00:00Z","html_url":"https://github.com/petry-projects/.github-private/actions/runs/99998","run_number":98}]\n'
+  exit 0
+fi
+exit 0
+STUBEOF
+  chmod +x "$STUB_DIR/gh"
+}
+
+# ---------------------------------------------------------------------------
+# Rate limit → exit 0, warning annotation
+# ---------------------------------------------------------------------------
+
+@test "Claude rate limit exits 0 and emits ::warning:: annotation" {
+  make_gh_stub_one_failure
+
+  cat > "$STUB_DIR/claude" << 'STUBEOF'
+#!/usr/bin/env bash
+# Simulate the exact rate-limit message seen in the failing run (2026-06-09)
+printf "You've hit your limit · resets 4pm (UTC)\n"
+exit 1
+STUBEOF
+  chmod +x "$STUB_DIR/claude"
+
+  run bash "$SCRIPT"
+
+  [ "$status" -eq 0 ]
+  [[ "$output" == *"::warning::"* ]]
+}
+
+@test "Claude rate limit: HAS_FAILURES is overridden to false in GITHUB_ENV" {
+  make_gh_stub_one_failure
+
+  cat > "$STUB_DIR/claude" << 'STUBEOF'
+#!/usr/bin/env bash
+printf "You've hit your limit · resets 4pm (UTC)\n"
+exit 1
+STUBEOF
+  chmod +x "$STUB_DIR/claude"
+
+  run bash "$SCRIPT"
+
+  [ "$status" -eq 0 ]
+  # HAS_FAILURES=false must appear after HAS_FAILURES=true so the last value wins
+  grep -q 'HAS_FAILURES=false' "$GITHUB_ENV"
+}
+
+@test "Claude 'usage limit' message also triggers rate-limit handling (exit 0)" {
+  make_gh_stub_one_failure
+
+  cat > "$STUB_DIR/claude" << 'STUBEOF'
+#!/usr/bin/env bash
+printf "usage limit reached for your account\n"
+exit 1
+STUBEOF
+  chmod +x "$STUB_DIR/claude"
+
+  run bash "$SCRIPT"
+
+  [ "$status" -eq 0 ]
+  [[ "$output" == *"::warning::"* ]]
+}
+
+# ---------------------------------------------------------------------------
+# Non-rate-limit Claude error → exit 1, error annotation
+# ---------------------------------------------------------------------------
+
+@test "non-rate-limit Claude failure exits 1 and emits ::error:: annotation" {
+  make_gh_stub_one_failure
+
+  cat > "$STUB_DIR/claude" << 'STUBEOF'
+#!/usr/bin/env bash
+printf "fatal: segmentation fault in model inference\n"
+exit 1
+STUBEOF
+  chmod +x "$STUB_DIR/claude"
+
+  run bash "$SCRIPT"
+
+  [ "$status" -eq 1 ]
+  [[ "$output" == *"::error::"* ]]
+}
+
+# ---------------------------------------------------------------------------
+# No failures → exit 0, Claude never invoked
+# ---------------------------------------------------------------------------
+
+@test "no failed runs in window exits 0 without invoking Claude" {
+  make_gh_stub_no_failures
+
+  # Claude stub that fails loudly if invoked — verifies Claude is not called
+  cat > "$STUB_DIR/claude" << 'STUBEOF'
+#!/usr/bin/env bash
+printf "UNEXPECTED: claude was invoked when no failures present\n"
+exit 1
+STUBEOF
+  chmod +x "$STUB_DIR/claude"
+
+  run bash "$SCRIPT"
+
+  [ "$status" -eq 0 ]
+  [[ "$output" != *"UNEXPECTED"* ]]
+}

--- a/tests/test_pr_review_health.bats
+++ b/tests/test_pr_review_health.bats
@@ -114,7 +114,11 @@ STUBEOF
 
   [ "$status" -eq 0 ]
   # HAS_FAILURES=false must appear after HAS_FAILURES=true so the last value wins
-  grep -q 'HAS_FAILURES=false' "$GITHUB_ENV"
+  true_line="$(grep -n '^HAS_FAILURES=true$' "$GITHUB_ENV" | tail -1 | cut -d: -f1)"
+  false_line="$(grep -n '^HAS_FAILURES=false$' "$GITHUB_ENV" | tail -1 | cut -d: -f1)"
+  [[ -n "$true_line" ]]
+  [[ -n "$false_line" ]]
+  (( false_line > true_line ))
 }
 
 @test "Claude 'usage limit' message also triggers rate-limit handling (exit 0)" {

--- a/tests/test_pr_review_health.bats
+++ b/tests/test_pr_review_health.bats
@@ -38,7 +38,7 @@ make_gh_stub_one_failure() {
 #!/usr/bin/env bash
 args="$*"
 # Token access check (per_page=1) — succeed silently
-if printf '%s' "$args" | grep -q 'per_page=1[^0]'; then
+if printf '%s' "$args" | grep -qE 'per_page=1([^0-9]|$)'; then
   exit 0
 fi
 # Run listing (per_page=100) — return one failed run
@@ -61,12 +61,24 @@ STUBEOF
   chmod +x "$STUB_DIR/gh"
 }
 
+# make_claude_stub: writes a claude stub that prints a message and exits.
+make_claude_stub() {
+  local message="$1"
+  local exit_code="${2:-1}"
+  cat > "$STUB_DIR/claude" << STUBEOF
+#!/usr/bin/env bash
+printf '%s\n' "$message"
+exit "$exit_code"
+STUBEOF
+  chmod +x "$STUB_DIR/claude"
+}
+
 # make_gh_stub_no_failures: returns only successful runs so Claude is skipped.
 make_gh_stub_no_failures() {
   cat > "$STUB_DIR/gh" << 'STUBEOF'
 #!/usr/bin/env bash
 args="$*"
-if printf '%s' "$args" | grep -q 'per_page=1[^0]'; then
+if printf '%s' "$args" | grep -qE 'per_page=1([^0-9]|$)'; then
   exit 0
 fi
 if printf '%s' "$args" | grep -q 'per_page=100'; then
@@ -85,13 +97,8 @@ STUBEOF
 @test "Claude rate limit exits 0 and emits ::warning:: annotation" {
   make_gh_stub_one_failure
 
-  cat > "$STUB_DIR/claude" << 'STUBEOF'
-#!/usr/bin/env bash
-# Simulate the exact rate-limit message seen in the failing run (2026-06-09)
-printf "You've hit your limit · resets 4pm (UTC)\n"
-exit 1
-STUBEOF
-  chmod +x "$STUB_DIR/claude"
+  # Simulate the exact rate-limit message seen in the failing run (2026-06-09)
+  make_claude_stub "You've hit your limit · resets 4pm (UTC)"
 
   run bash "$SCRIPT"
 
@@ -101,13 +108,7 @@ STUBEOF
 
 @test "Claude rate limit: HAS_FAILURES is overridden to false in GITHUB_ENV" {
   make_gh_stub_one_failure
-
-  cat > "$STUB_DIR/claude" << 'STUBEOF'
-#!/usr/bin/env bash
-printf "You've hit your limit · resets 4pm (UTC)\n"
-exit 1
-STUBEOF
-  chmod +x "$STUB_DIR/claude"
+  make_claude_stub "You've hit your limit · resets 4pm (UTC)"
 
   run bash "$SCRIPT"
 
@@ -118,13 +119,7 @@ STUBEOF
 
 @test "Claude 'usage limit' message also triggers rate-limit handling (exit 0)" {
   make_gh_stub_one_failure
-
-  cat > "$STUB_DIR/claude" << 'STUBEOF'
-#!/usr/bin/env bash
-printf "usage limit reached for your account\n"
-exit 1
-STUBEOF
-  chmod +x "$STUB_DIR/claude"
+  make_claude_stub "usage limit reached for your account"
 
   run bash "$SCRIPT"
 
@@ -138,13 +133,7 @@ STUBEOF
 
 @test "non-rate-limit Claude failure exits 1 and emits ::error:: annotation" {
   make_gh_stub_one_failure
-
-  cat > "$STUB_DIR/claude" << 'STUBEOF'
-#!/usr/bin/env bash
-printf "fatal: segmentation fault in model inference\n"
-exit 1
-STUBEOF
-  chmod +x "$STUB_DIR/claude"
+  make_claude_stub "fatal: segmentation fault in model inference"
 
   run bash "$SCRIPT"
 
@@ -160,12 +149,7 @@ STUBEOF
   make_gh_stub_no_failures
 
   # Claude stub that fails loudly if invoked — verifies Claude is not called
-  cat > "$STUB_DIR/claude" << 'STUBEOF'
-#!/usr/bin/env bash
-printf "UNEXPECTED: claude was invoked when no failures present\n"
-exit 1
-STUBEOF
-  chmod +x "$STUB_DIR/claude"
+  make_claude_stub "UNEXPECTED: claude was invoked when no failures present"
 
   run bash "$SCRIPT"
 

